File: src/lib/producer/contracts/cut-repair-picture-plan-authority.ts
import type {
  CutRestoreSpeechV1,
  FrameRangeV1,
  SampleRangeV1,
} from "./cut-restore-speech-v1";
import type { ProjectRevision } from "./project-revision";
import {
  exactKeys,
  objectValue,
  sha256,
} from "./validation";
import { canonicalJsonSha256 } from
  "@/lib/server/auto-edit-hash";

const FIELD = "cutRepairPicturePlanAuthority";
const KEYS = [
  "schemaVersion", "kind", "operationHash", "parentPlanObjectHash",
  "parentTimelineMapHash", "childTimelineMapHash", "target",
  "parentCutRow", "parentCutRowHash", "childCutRow", "childCutRowHash",
  "sourceExtension", "reclaimedSilence", "dirtyFrameRange",
  "mappingProof", "mappingProofHash", "reversion", "authorityHash",
] as const;
const TARGET_KEYS = [
  "index", "segmentId", "parentElementVersion", "childElementVersion",
] as const;
const REVERSION_KEYS = [
  "action", "index", "segmentId", "expectedChildCutRowHash",
  "restoreParentCutRowHash",
] as const;

export interface CutRepairPicturePlanAuthorityV1 {
  schemaVersion: 1;
  kind: "cut-repair-picture-plan-authority";
  operationHash: string;
  parentPlanObjectHash: string;
  parentTimelineMapHash: string;
  childTimelineMapHash: string;
  target: {
    index: number;
    segmentId: string;
    parentElementVersion: number;
    childElementVersion: number;
  };
  parentCutRow: Record<string, unknown>;
  parentCutRowHash: string;
  childCutRow: Record<string, unknown>;
  childCutRowHash: string;
  sourceExtension: SampleRangeV1;
  reclaimedSilence: SampleRangeV1;
  dirtyFrameRange: FrameRangeV1;
  mappingProof: Record<string, unknown>;
  mappingProofHash: string;
  reversion: Record<string, unknown>;
  authorityHash: string;
}

function integer(value: unknown, minimum: number, label: string): number {
  if (!Number.isSafeInteger(value) || Number(value) < minimum) {
    throw new Error(`${label} must be a safe integer >= ${minimum}`);
  }
  return Number(value);
}

function frameRange(value: unknown, label: string): FrameRangeV1 {
  const row = objectValue(value, label);
  const keys = ["startFrame", "endFrameExclusive"] as const;
  exactKeys(row, keys, keys, label);
  const startFrame = integer(row.startFrame, 0, `${label}.startFrame`);
  const endFrameExclusive = integer(
    row.endFrameExclusive, 1, `${label}.endFrameExclusive`);
  if (endFrameExclusive <= startFrame) throw new Error(`${label} is empty`);
  return { startFrame, endFrameExclusive };
}

function sampleRange(value: unknown, label: string): SampleRangeV1 {
  const row = objectValue(value, label);
  const keys = ["startSample", "endSampleExclusive"] as const;
  exactKeys(row, keys, keys, label);
  const startSample = integer(row.startSample, 0, `${label}.startSample`);
  const endSampleExclusive = integer(
    row.endSampleExclusive, 1, `${label}.endSampleExclusive`);
  if (endSampleExclusive <= startSample) throw new Error(`${label} is empty`);
  return { startSample, endSampleExclusive };
}

function elementVersion(
  row: Record<string, unknown>,
  label: string,
): number {
  if (row.generation !== undefined && row.version !== undefined) {
    throw new Error(`${label} has two version fields`);
  }
  return integer(row.generation ?? row.version ?? 1, 1, `${label} version`);
}

function parseTarget(value: unknown) {
  const row = objectValue(value, "picture plan target");
  exactKeys(row, TARGET_KEYS, TARGET_KEYS, "picture plan target");
  if (typeof row.segmentId !== "string" || !row.segmentId) {
    throw new Error("picture plan target segmentId is empty");
  }
  return {
    index: integer(row.index, 0, "picture plan target index"),
    segmentId: row.segmentId,
    parentElementVersion: integer(
      row.parentElementVersion, 1, "picture plan parent version"),
    childElementVersion: integer(
      row.childElementVersion, 1, "picture plan child version"),
  };
}

function parseReversion(value: unknown): Record<string, unknown> {
  const row = objectValue(value, "picture plan reversion");
  exactKeys(row, REVERSION_KEYS, REVERSION_KEYS, "picture plan reversion");
  if (row.action !== "restore-parent-cut-row"
      || typeof row.segmentId !== "string" || !row.segmentId) {
    throw new Error("picture plan reversion is unsupported");
  }
  integer(row.index, 0, "picture plan reversion index");
  sha256(row.expectedChildCutRowHash, "picture plan reversion child");
  sha256(row.restoreParentCutRowHash, "picture plan reversion parent");
  return row;
}

function parseMappingProof(
  value: unknown,
  expectedHash: string,
): Record<string, unknown> {
  const row = objectValue(value, "picture plan mapping proof");
  const keys = [
    "schemaVersion", "parentTimelineMapHash", "childTimelineMapHash",
    "authorizedDirtyWindows", "unchangedRanges",
  ] as const;
  exactKeys(row, keys, keys, "picture plan mapping proof");
  if (row.schemaVersion !== 1
      || !Array.isArray(row.authorizedDirtyWindows)
      || !row.authorizedDirtyWindows.length
      || !Array.isArray(row.unchangedRanges)
      || !row.unchangedRanges.length
      || canonicalJsonSha256(row) !== expectedHash) {
    throw new Error("picture plan mapping proof is stale");
  }
  row.authorizedDirtyWindows.forEach((item, index) =>
    frameRange(item, `mapping authorizedDirtyWindows[${index}]`));
  return row;
}

function assertRowBindings(
  authority: CutRepairPicturePlanAuthorityV1,
): void {
  const target = authority.target;
  const parentVersion = elementVersion(
    authority.parentCutRow, "picture plan parent row");
  const childVersion = elementVersion(
    authority.childCutRow, "picture plan child row");
  const reversion = authority.reversion;
  if (authority.parentCutRow.id !== target.segmentId
      || authority.childCutRow.id !== target.segmentId
      || parentVersion !== target.parentElementVersion
      || childVersion !== target.childElementVersion
      || childVersion !== parentVersion + 1
      || reversion.index !== target.index
      || reversion.segmentId !== target.segmentId
      || reversion.expectedChildCutRowHash !== authority.childCutRowHash
      || reversion.restoreParentCutRowHash !== authority.parentCutRowHash) {
    throw new Error("picture plan row or reversion binding is stale");
  }
}

/** Parse the plan-carried exact row mutation and deterministic reversion. */
export function parseCutRepairPicturePlanAuthorityV1(
  value: unknown,
): CutRepairPicturePlanAuthorityV1 {
  const row = objectValue(value, "CutRepairPicturePlanAuthorityV1");
  exactKeys(row, KEYS, KEYS, "CutRepairPicturePlanAuthorityV1");
  const core = Object.fromEntries(
    Object.entries(row).filter(([key]) => key !== "authorityHash"));
  if (row.schemaVersion !== 1
      || row.kind !== "cut-repair-picture-plan-authority"
      || canonicalJsonSha256(core) !== row.authorityHash) {
    throw new Error("picture plan authority is stale");
  }
  const parentCutRow = objectValue(row.parentCutRow, "picture plan parent row");
  const childCutRow = objectValue(row.childCutRow, "picture plan child row");
  const parentCutRowHash = sha256(
    row.parentCutRowHash, "picture plan parent row hash");
  const childCutRowHash = sha256(
    row.childCutRowHash, "picture plan child row hash");
  if (canonicalJsonSha256(parentCutRow) !== parentCutRowHash
      || canonicalJsonSha256(childCutRow) !== childCutRowHash) {
    throw new Error("picture plan cut-row hash is stale");
  }
  const mappingProofHash = sha256(
    row.mappingProofHash, "picture plan mapping proof hash");
  const authority = {
    schemaVersion: 1,
    kind: "cut-repair-picture-plan-authority",
    operationHash: sha256(row.operationHash, "picture plan operation"),
    parentPlanObjectHash: sha256(
      row.parentPlanObjectHash, "picture plan parent plan"),
    parentTimelineMapHash: sha256(
      row.parentTimelineMapHash, "picture plan parent timeline"),
    childTimelineMapHash: sha256(
      row.childTimelineMapHash, "picture plan child timeline"),
    target: parseTarget(row.target),
    parentCutRow, parentCutRowHash, childCutRow, childCutRowHash,
    sourceExtension: sampleRange(
      row.sourceExtension, "picture plan source extension"),
    reclaimedSilence: sampleRange(
      row.reclaimedSilence, "picture plan reclaimed silence"),
    dirtyFrameRange: frameRange(
      row.dirtyFrameRange, "picture plan dirty frame range"),
    mappingProof: parseMappingProof(row.mappingProof, mappingProofHash),
    mappingProofHash,
    reversion: parseReversion(row.reversion),
    authorityHash: sha256(row.authorityHash, "picture plan authority"),
  } satisfies CutRepairPicturePlanAuthorityV1;
  assertRowBindings(authority);
  return authority;
}

function assertOperationBinding(
  value: CutRepairPicturePlanAuthorityV1,
  operation: CutRestoreSpeechV1,
): void {
  const dirty = operation.pictureDirtyWindows;
  const reclaimed = operation.reclaimedSilence;
  if (operation.method !== "extend-and-reclaim-silence"
      || dirty.length !== 1 || !reclaimed
      || value.operationHash !== canonicalJsonSha256(operation)
      || value.parentTimelineMapHash !== operation.parentTimelineMapHash
      || value.target.segmentId !== operation.segment.segmentId
      || value.target.parentElementVersion !== operation.segment.elementVersion
      || canonicalJsonSha256(value.sourceExtension)
        !== canonicalJsonSha256(operation.sourceExtension)
      || canonicalJsonSha256(value.reclaimedSilence)
        !== canonicalJsonSha256(reclaimed.sourceSampleRange)
      || canonicalJsonSha256(value.dirtyFrameRange)
        !== canonicalJsonSha256(dirty[0])) {
    throw new Error("picture plan authority contradicts the repair operation");
  }
}

/**
 * Require picture operations to carry one exact plan authority, while
 * forbidding an audio-only plan from manufacturing picture authority.
 */
export function bindCutRepairPicturePlanAuthority(
  plan: Record<string, unknown>,
  operation: CutRestoreSpeechV1,
  childTimelineMapHash: string,
): CutRepairPicturePlanAuthorityV1 | null {
  const raw = plan[FIELD];
  if (!operation.pictureDirtyWindows.length) {
    if (raw !== undefined) {
      throw new Error("audio-only repair plan carries picture authority");
    }
    return null;
  }
  if (raw === undefined) {
    throw new Error("picture repair plan lacks picture-plan authority");
  }
  const authority = parseCutRepairPicturePlanAuthorityV1(raw);
  assertOperationBinding(authority, operation);
  const track = plan.cutTrack;
  const index = authority.target.index;
  if (!Array.isArray(track) || index >= track.length
      || canonicalJsonSha256(track[index])
        !== canonicalJsonSha256(authority.childCutRow)
      || authority.childTimelineMapHash
        !== sha256(childTimelineMapHash, "picture plan child timeline")
      || authority.mappingProof.parentTimelineMapHash
        !== authority.parentTimelineMapHash
      || authority.mappingProof.childTimelineMapHash
        !== authority.childTimelineMapHash
      || canonicalJsonSha256(authority.mappingProof.authorizedDirtyWindows)
        !== canonicalJsonSha256([authority.dirtyFrameRange])) {
    throw new Error("picture plan authority does not bind the review plan");
  }
  return authority;
}

/** Bind the reopened plan mutation to its exact V2 parent revision. */
export function assertCutRepairPicturePlanParent(
  authority: CutRepairPicturePlanAuthorityV1 | null,
  operation: CutRestoreSpeechV1,
  parent: ProjectRevision,
): void {
  if (!authority) return;
  if (parent.schemaVersion !== 2
      || authority.parentPlanObjectHash !== parent.planObjectHash
      || authority.parentTimelineMapHash !== parent.timelineMapHash
      || operation.parentPictureLockHash !== parent.pictureLockHash) {
    throw new Error("picture repair plan does not bind parent authority");
  }
}

File: scripts/producer/edit/cut_repair_picture_plan_authority.py
"""Closed mutation and reversion authority for a picture review plan."""
from __future__ import annotations

from dataclasses import dataclass

from edit.picture_lock_common import content_hash

PLAN_FIELD = "cutRepairPicturePlanAuthority"
_HEX = frozenset("0123456789abcdef")
_KEYS = {
    "schemaVersion", "kind", "operationHash", "parentPlanObjectHash",
    "parentTimelineMapHash", "childTimelineMapHash", "target",
    "parentCutRow", "parentCutRowHash", "childCutRow", "childCutRowHash",
    "sourceExtension", "reclaimedSilence", "dirtyFrameRange",
    "mappingProof", "mappingProofHash", "reversion", "authorityHash",
}
_TARGET_KEYS = {
    "index", "segmentId", "parentElementVersion", "childElementVersion",
}
_REVERSION_KEYS = {
    "action", "index", "segmentId", "expectedChildCutRowHash",
    "restoreParentCutRowHash",
}


class PicturePlanAuthorityError(ValueError):
    """A picture plan mutation receipt is open or stale."""


def _sha256(value: object) -> bool:
    return isinstance(value, str) and len(value) == 64 \
        and all(char in _HEX for char in value)


def _element_version(row: object) -> int | None:
    if not isinstance(row, dict) \
            or "generation" in row and "version" in row:
        return None
    value = row.get("generation", row.get("version", 1))
    return value if type(value) is int and value > 0 else None


@dataclass(frozen=True)
class PicturePlanAuthorityInput:
    """Exact facts proved by the picture review-plan builder."""

    operation_hash: str
    parent_plan_hash: str
    index: int
    parent_row: dict
    parent_version: int
    child_row: dict
    child_version: int
    source_extension: dict
    reclaimed_silence: dict
    dirty_frame_range: dict
    parent_map_hash: str
    child_map_hash: str
    mapping_proof: dict
    mapping_proof_hash: str


def _target(value: PicturePlanAuthorityInput) -> dict:
    return {
        "index": value.index,
        "segmentId": value.parent_row["id"],
        "parentElementVersion": value.parent_version,
        "childElementVersion": value.child_version,
    }


def _reversion(value: PicturePlanAuthorityInput) -> dict:
    return {
        "action": "restore-parent-cut-row",
        "index": value.index,
        "segmentId": value.parent_row["id"],
        "expectedChildCutRowHash": content_hash(value.child_row),
        "restoreParentCutRowHash": content_hash(value.parent_row),
    }


def build_picture_plan_authority(
    value: PicturePlanAuthorityInput,
) -> dict:
    """Seal the exact row replacement and its deterministic reversion."""
    core = {
        "schemaVersion": 1,
        "kind": "cut-repair-picture-plan-authority",
        "operationHash": value.operation_hash,
        "parentPlanObjectHash": value.parent_plan_hash,
        "parentTimelineMapHash": value.parent_map_hash,
        "childTimelineMapHash": value.child_map_hash,
        "target": _target(value),
        "parentCutRow": value.parent_row,
        "parentCutRowHash": content_hash(value.parent_row),
        "childCutRow": value.child_row,
        "childCutRowHash": content_hash(value.child_row),
        "sourceExtension": value.source_extension,
        "reclaimedSilence": value.reclaimed_silence,
        "dirtyFrameRange": value.dirty_frame_range,
        "mappingProof": value.mapping_proof,
        "mappingProofHash": value.mapping_proof_hash,
        "reversion": _reversion(value),
    }
    return parse_picture_plan_authority({
        **core, "authorityHash": content_hash(core)})


def _closed_rows(core: dict) -> bool:
    target = core.get("target")
    reversion = core.get("reversion")
    return (
        isinstance(target, dict) and set(target) == _TARGET_KEYS
        and isinstance(reversion, dict) and set(reversion) == _REVERSION_KEYS
    )


def _hash_bindings(core: dict) -> bool:
    hashes = (
        "operationHash", "parentPlanObjectHash", "parentTimelineMapHash",
        "childTimelineMapHash", "parentCutRowHash", "childCutRowHash",
        "mappingProofHash",
    )
    return (
        all(_sha256(core.get(key)) for key in hashes)
        and core.get("parentCutRowHash")
        == content_hash(core.get("parentCutRow"))
        and core.get("childCutRowHash")
        == content_hash(core.get("childCutRow"))
        and core.get("mappingProofHash")
        == content_hash(core.get("mappingProof"))
    )


def _target_binding(core: dict) -> bool:
    target = core["target"]
    parent, child = core.get("parentCutRow"), core.get("childCutRow")
    parent_version = _element_version(parent)
    child_version = _element_version(child)
    return (
        type(target.get("index")) is int and target["index"] >= 0
        and isinstance(target.get("segmentId"), str)
        and bool(target["segmentId"])
        and isinstance(parent, dict) and isinstance(child, dict)
        and parent.get("id") == child.get("id") == target["segmentId"]
        and parent_version == target.get("parentElementVersion")
        and child_version == target.get("childElementVersion")
        and parent_version is not None
        and child_version == parent_version + 1
    )


def _reversion_binding(core: dict) -> bool:
    target = core["target"]
    reversion = core["reversion"]
    return (
        reversion.get("action") == "restore-parent-cut-row"
        and reversion.get("index") == target.get("index")
        and reversion.get("segmentId") == target.get("segmentId")
        and reversion.get("expectedChildCutRowHash")
        == core.get("childCutRowHash")
        and reversion.get("restoreParentCutRowHash")
        == core.get("parentCutRowHash")
    )


def parse_picture_plan_authority(value: object) -> dict:
    """Reopen one exact plan-carried mutation/reversion receipt."""
    if not isinstance(value, dict) or set(value) != _KEYS:
        raise PicturePlanAuthorityError("picture plan authority is not closed")
    core = {key: item for key, item in value.items()
            if key != "authorityHash"}
    if core.get("schemaVersion") != 1 \
            or core.get("kind") != "cut-repair-picture-plan-authority" \
            or not _closed_rows(core) \
            or not _hash_bindings(core) \
            or not _target_binding(core) \
            or not _reversion_binding(core) \
            or not _sha256(value.get("authorityHash")) \
            or value.get("authorityHash") != content_hash(core):
        raise PicturePlanAuthorityError("picture plan authority is stale")
    return dict(value)


def validate_picture_plan_authority_for_plan(
    plan: object,
) -> dict | None:
    """Require a plan-carried receipt to bind its exact child row and map."""
    if not isinstance(plan, dict):
        raise PicturePlanAuthorityError("picture plan must be an object")
    value = plan.get(PLAN_FIELD)
    if value is None:
        return None
    authority = parse_picture_plan_authority(value)
    target = authority["target"]
    track = plan.get("cutTrack")
    index = target["index"]
    if not isinstance(track, list) or index >= len(track) \
            or track[index] != authority["childCutRow"]:
        raise PicturePlanAuthorityError(
            "picture plan authority does not bind cutTrack")
    from edit.compatibility_projection import build_projection, stable_digest
    projection = build_projection(plan, stable_digest(plan))
    if projection["timelineMapHash"] != authority["childTimelineMapHash"]:
        raise PicturePlanAuthorityError(
            "picture plan authority does not bind child timeline")
    return authority
